crates/linker/src/pe.rs

```diff
@@ -1,47 +1,343 @@
-use std::{ops::Range, path::Path, time::Instant};
+use std::{
+    io::{BufReader, BufWriter},
+    ops::Range,
+    path::Path,
+    time::Instant,
+};
 
-use memmap2::{Mmap, MmapMut};
+use bincode::{deserialize_from, serialize_into};
+use memmap2::MmapMut;
 use object::{
-    pe::{ImageImportDescriptor, ImageNtHeaders64, ImageSectionHeader, ImageThunkData64},
+    pe::{
+        self, ImageFileHeader, ImageImportDescriptor, ImageNtHeaders64, ImageSectionHeader,
+        ImageThunkData64,
+    },
     read::pe::ImportTable,
-    LittleEndian as LE, RelocationTarget, SectionIndex,
+    LittleEndian as LE, Object, RelocationTarget, SectionIndex,
 };
+use serde::{Deserialize, Serialize};
+
 use roc_collections::MutMap;
 use roc_error_macros::internal_error;
 
-use crate::{load_struct_inplace, load_struct_inplace_mut};
+use crate::{generate_dylib::APP_DLL, load_struct_inplace, load_struct_inplace_mut};
+
+#[derive(Debug, Serialize, Deserialize)]
+struct PeMetadata {
+    dynhost_file_size: usize,
+
+    last_host_section_size: u64,
+    last_host_section_address: u64,
+
+    optional_header_offset: usize,
+
+    dynamic_relocations: DynamicRelocationsPe,
+
+    /// File offset for the thunks of our dummy .dll
+    thunks_start_offset: usize,
+
+    /// Constants from the host .exe header
+    image_base: u64,
+    file_alignment: u32,
+    section_alignment: u32,
+
+    /// Symbols that the host imports, like roc__mainForHost_1_exposed_generic
+    imports: Vec<String>,
+
+    /// Symbols that the host exports, like roc_alloc
+    exports: MutMap<String, i64>,
+}
 
 #[allow(dead_code)]
 pub(crate) fn preprocess_windows(
-    exec_filename: &str,
-    _metadata_filename: &str,
-    _out_filename: &str,
+    host_exe_filename: &str,
+    metadata_filename: &Path,
+    out_filename: &Path,
     _shared_lib: &Path,
     _verbose: bool,
     _time: bool,
 ) -> object::read::Result<()> {
+    use object::ObjectSection;
+
     let total_start = Instant::now();
     let exec_parsing_start = total_start;
-    let exec_file = std::fs::File::open(exec_filename).unwrap_or_else(|e| internal_error!("{}", e));
-    let exec_mmap = unsafe { Mmap::map(&exec_file).unwrap_or_else(|e| internal_error!("{}", e)) };
-    let exec_data = &*exec_mmap;
-    let _exec_obj = match object::read::pe::PeFile64::parse(exec_data) {
+
+    let _exec_parsing_duration = exec_parsing_start.elapsed();
```

Comment:
Are you planning to use this timing stuff in a later PR?

Reply:
it seems useful. Our claim is that we are much faster than normal linking, and I'd like to be able to back that up with data